File: frontend/src/lib/api/operation.ts
// 프론트 B 담당 — 운영 API (백엔드 /api/v1/operation 연동)
// 백엔드 응답은 CommonResponse { success, data, message } 래핑 → unwrap 해서 data만 반환
import { apiFetch } from './client';

const auth = (token: string) => ({ Authorization: `Bearer ${token}` });

type CommonResponse<T> = { success: boolean; data: T; message: string };

/** CommonResponse에서 data만 꺼내되, success=false면 message로 에러 throw */
function unwrap<T>(res: CommonResponse<T>): T {
  if (!res || res.success === false) {
    throw new Error(res?.message ?? '요청이 실패했습니다.');
  }
  return res.data;
}

// ---------- 타입 ----------
export type TaxLine = { name: string; amount: number; basis: string };
export type TaxFiling = {
  name: string;
  due_date: string;
  dday: number;
  status: string;
  note: string;
};
export type TaxEstimate = {
  period: string;
  tax_type: string;
  total_revenue: number;
  total_expense: number;
  taxable_base: number;
  vat: number;
  income_tax: number;
  withholding_tax: number;
  total_tax: number;
  lines: TaxLine[];
  filing_schedule: TaxFiling[];
  next_filing: TaxFiling | null;
  summary: string;
  disclaimer: string;
};

export type Settlement = {
  total_sales: number;
  total_expense: number;
  total_payroll: number;
  net_profit: number;
  other_expense?: number;
  year_month?: string;
  period_start?: string;
  period_end?: string;
  calculated_at?: string;
  disclaimer?: string;
};

export type Forecast = {
  target_date: string;
  predicted_sales: number;
  predicted_quantity: number;
  engine: string;
  evidence_summary: string;
};

export type Expense = {
  id: number;
  store_id: string;
  amount: number;
  category: string;
  description: string | null;
  expense_date: string;
  created_at: string;
};

export type Payroll = {
  employee_id: number;
  employee_name: string;
  role: string;
  hourly_rate: number;
  period_start?: string;
  period_end?: string;
  total_work_hours: number;
  base_salary: number;
  weekly_holiday_allowance: number;
  estimated_salary: number;
  based_on_actual?: boolean;
};

// ---------- 세무 ----------
/** 세무 예상 계산 — DB 매출·비용 자동집계 (인증 필요, 로그인 매장 기준) */
export async function getTaxEstimate(
  token: string,
  yearMonth: string,
  taxType: 'general' | 'simplified' = 'general',
): Promise<TaxEstimate> {
  return unwrap(
    await apiFetch<CommonResponse<TaxEstimate>>(
      `/api/v1/operation/tax/estimate?year_month=${yearMonth}&tax_type=${taxType}`,
      { headers: auth(token) },
    ),
  );
}

/** 세무 예상 계산 — 매출·비용 직접 입력 (인증 불필요) */
export async function estimateTaxManual(body: {
  period: string;
  total_revenue: number;
  total_expense: number;
  tax_type?: 'general' | 'simplified';
}): Promise<TaxEstimate> {
  return unwrap(
    await apiFetch<CommonResponse<TaxEstimate>>('/api/v1/operation/tax/estimate', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  );
}

// ---------- 정산 ----------
/** 월별 손익 정산 (매출−비용−인건비). 연월을 기간으로 변환해 계산 API 호출 */
export async function getSettlement(yearMonth: string): Promise<Settlement> {
  const [y, m] = yearMonth.split('-').map(Number);
  const lastDay = new Date(y, m, 0).getDate();
  const data = unwrap(
    await apiFetch<CommonResponse<Settlement | Settlement[]>>(
      '/api/v1/operation/settlements/calculate',
      {
        method: 'POST',
        body: JSON.stringify({
          period_start: `${yearMonth}-01`,
          period_end: `${yearMonth}-${String(lastDay).padStart(2, '0')}`,
        }),
      },
    ),
  );
  const item = Array.isArray(data) ? data[0] : data;
  if (!item) throw new Error('정산 데이터가 없습니다.');
  return item;
}

// ---------- 판매예측 ----------
/** 판매 예측 (ARIMA + DB 자동집계, 데이터 부족 시 이동평균 폴백) */
export async function forecastSales(body: {
  target_date: string;
  store_id?: string;
  has_event?: boolean;
  engine?: 'arima' | 'average';
}): Promise<Forecast> {
  return unwrap(
    await apiFetch<CommonResponse<Forecast>>('/api/v1/operation/forecast/sales', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  );
}

// ---------- 지출 ----------
/** 지출 등록 (정산·세무의 비용 데이터 소스) */
export async function createExpense(
  token: string,
  body: { amount: number; category: string; expense_date: string; description?: string },
): Promise<Expense> {
  return unwrap(
    await apiFetch<CommonResponse<Expense>>('/api/v1/operation/expenses', {
      method: 'POST',
      headers: auth(token),
      body: JSON.stringify(body),
    }),
  );
}

/** 지출 목록 조회 */
export async function listExpenses(token: string, yearMonth?: string): Promise<Expense[]> {
  const q = yearMonth ? `?year_month=${yearMonth}` : '';
  return unwrap(
    await apiFetch<CommonResponse<Expense[]>>(`/api/v1/operation/expenses${q}`, {
      headers: auth(token),
    }),
  );
}

// ---------- 급여 ----------
/** 등록된 모든 직원의 해당 월 예상 급여 목록 (실근무 우선, 없으면 계획시간 기준) */
export async function listPayroll(yearMonth: string): Promise<Payroll[]> {
  return unwrap(
    await apiFetch<CommonResponse<Payroll[]>>(`/api/v1/operation/payroll/all?year_month=${yearMonth}`),
  );
}

// ---------- 근무 스케줄 ----------
export type Schedule = {
  id: number;
  employee_id: number;
  start_time: string;
  end_time: string;
  date: string; // YYYY-MM-DD
  actual_start_time?: string | null;
  actual_end_time?: string | null;
};

/** 등록된 근무 스케줄 전체 조회 */
export async function listSchedules(): Promise<Schedule[]> {
  return unwrap(await apiFetch<CommonResponse<Schedule[]>>('/api/v1/operation/schedules'));
}

/** 근무 스케줄 등록 */
export async function createSchedule(body: {
  employee_id: number;
  start_time: string;
  end_time: string;
}): Promise<Schedule> {
  return unwrap(
    await apiFetch<CommonResponse<Schedule>>('/api/v1/operation/schedules', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  );
}

/** 근무 스케줄 시간 수정 */
export async function updateSchedule(
  id: number,
  body: { start_time?: string; end_time?: string },
): Promise<Schedule> {
  return unwrap(
    await apiFetch<CommonResponse<Schedule>>(`/api/v1/operation/schedules/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),
  );
}

/** 근무 스케줄 삭제 */
export async function deleteSchedule(id: number): Promise<null> {
  return unwrap(
    await apiFetch<CommonResponse<null>>(`/api/v1/operation/schedules/${id}`, { method: 'DELETE' }),
  );
}

// ---------- 챗봇 / ERP 신규: 직원별 기피/불가 시간 & 스케줄 추천 ----------
export type EmployeeUnavailability = {
  id: number;
  employee_id: number;
  employee_name?: string;
  unavailability_type: 'weekly_recurring' | 'specific_date';
  day_of_week?: number;
  specific_date?: string;
  start_hour: number;
  end_hour: number;
  restriction_level: 'hard' | 'soft';
  reason?: string;
  created_at: string;
};

export type AssignedEmployee = {
  id: number;
  name: string;
  role: string;
  level: 'hard' | 'soft' | null;
};

export type HourlyRecommendation = {
  hour: number;
  predicted_sales: number;
  predicted_profit: number;
  recommended_employee_count: number;
  busy_level: 'PEAK' | 'HIGH' | 'NORMAL' | 'LOW';
  assigned_employees: AssignedEmployee[];
  unassigned_count: number;
};

export type ScheduleRecommendation = {
  target_date: string;
  hourly_recommendations?: HourlyRecommendation[];
  total_recommended_hours: number;
  estimated_payroll_cost: number;
  warnings?: string[];
  summary: string;
};

/** 직원 기피/불가 시간 등록 API */
export async function createUnavailability(
  token: string,
  body: {
    employee_id: number;
    unavailability_type: 'weekly_recurring' | 'specific_date';
    day_of_week?: number;
    specific_date?: string;
    start_hour: number;
    end_hour: number;
    restriction_level: 'hard' | 'soft';
    reason?: string;
  },
): Promise<EmployeeUnavailability> {
  return unwrap(
    await apiFetch<CommonResponse<EmployeeUnavailability>>('/api/v1/operation/unavailability', {
      method: 'POST',
      headers: auth(token),
      body: JSON.stringify(body),
    }),
  );
}

/** 직원 기피/불가 시간 목록 조회 API */
export async function listUnavailabilities(
  token: string,
  employeeId?: number,
): Promise<EmployeeUnavailability[]> {
  const q = employeeId ? `?employee_id=${employeeId}` : '';
  return unwrap(
    await apiFetch<CommonResponse<EmployeeUnavailability[]>>(`/api/v1/operation/unavailability${q}`, {
      headers: auth(token),
    }),
  );
}

/** 직원 기피/불가 시간 삭제 API */
export async function deleteUnavailability(token: string, unavailabilityId: number): Promise<void> {
  await apiFetch<CommonResponse<null>>(`/api/v1/operation/unavailability/${unavailabilityId}`, {
    method: 'DELETE',
    headers: auth(token),
  });
}

/** 알바 스케줄 추천 API (기피시간 반영) */
export async function getScheduleRecommendation(
  token: string,
  targetDate: string,
  storeId: string = 'store_gildong',
): Promise<ScheduleRecommendation> {
  return unwrap(
    await apiFetch<CommonResponse<ScheduleRecommendation>>('/api/v1/operation/schedules/recommend', {
      method: 'POST',
      headers: auth(token),
      body: JSON.stringify({ target_date: targetDate, store_id: storeId }),
    }),
  );
}

/** AI 스케줄 추천 — 과거 매출 시간대 분석 기반 */
export async function recommendSchedule(body: {
  target_date: string;
  store_id?: string;
}): Promise<ScheduleRecommendation> {
  return unwrap(
    await apiFetch<CommonResponse<ScheduleRecommendation>>('/api/v1/operation/schedules/recommend', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  );
}

File: backend/app/models/operation.py
# c:\STUDY\SimpleM\backend\app\models\operation.py
"""운영/예측 모델 (백엔드 C) - 백엔드 A가 SQLAlchemy 표준 ORM 규격으로 수정"""
import logging
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Date, Float, inspect, text
from sqlalchemy.sql import func
from app.core.database import Base

logger = logging.getLogger(__name__)

# 1. 직원 정보를 저장할 데이터베이스 테이블 설계도입니다.
class Employee(Base):
    """직원 정보 모델"""
    __tablename__ = "employees"  # 데이터베이스 내의 실제 테이블 명

    id = Column(Integer, primary_key=True, index=True)
    store_id = Column(String(100), nullable=True, index=True)  # 소속 매장(점주 이메일). 정산·급여 매장별 스코핑용
    name = Column(String(50), nullable=False)      # 직원 이름
    hourly_rate = Column(Integer, nullable=False)   # 시급 (KRW)
    role = Column(String(50), nullable=False)       # 직책/역할 (예: 바리스타)


def ensure_employee_store_column(engine) -> None:
    """[자가치유 스키마] 기존 employees 테이블에 store_id 컬럼이 없으면 멱등하게 추가한다.
    create_all은 기존 테이블을 ALTER하지 않으므로 배포 시 무중단으로 보강.
    기존(매장 미지정) 직원은 데모 기본 매장(owner@cafe.com)으로 백필해 정산 누락을 막는다."""
    try:
        insp = inspect(engine)
        if not insp.has_table("employees"):
            return
        existing = {c["name"] for c in insp.get_columns("employees")}
    except Exception as e:
        logger.warning(f"[직원 스키마] employees 점검 실패 — 건너뜁니다: {e}")
        return
    if "store_id" in existing:
        return
    try:
        with engine.begin() as conn:
            conn.execute(text("ALTER TABLE employees ADD COLUMN store_id VARCHAR(100)"))
            # 기존 직원은 데모 기본 매장으로 귀속 (매장별 정산에서 누락되지 않게)
            conn.execute(text("UPDATE employees SET store_id = 'owner@cafe.com' WHERE store_id IS NULL"))
        logger.info("[직원 스키마] employees.store_id 컬럼 추가 + 기존 직원 백필 완료")
    except Exception as e:
        logger.warning(f"[직원 스키마] store_id 보강 실패: {e}")


# 2. 알바생들의 스케줄을 저장할 데이터베이스 테이블 설계도입니다.
class Schedule(Base):
    """근무 스케줄 모델"""
    __tablename__ = "schedules"  # 데이터베이스 내의 실제 테이블 명

    id = Column(Integer, primary_key=True, index=True)
    
    # 이 스케줄이 어떤 직원(Employee)을 뜻하는지 외래키로 연결합니다.
    employee_id = Column(Integer, ForeignKey("employees.id", ondelete="CASCADE"), nullable=False)
    
    start_time = Column(DateTime, nullable=False)  # 근무 시작 시간
    end_time = Column(DateTime, nullable=False)    # 근무 종료 시간
    date = Column(String(20), nullable=False)      # 근무 일자 (YYYY-MM-DD)
    actual_start_time = Column(DateTime, nullable=True)  # 실제 출근 시간 (신규 추가)
    actual_end_time = Column(DateTime, nullable=True)    # 실제 퇴근 시간 (신규 추가)


# 3. 매장의 매입 비용 등 지출을 저장할 데이터베이스 테이블 설계도입니다.
class Expense(Base):
    """지출 비용 모델"""
    __tablename__ = "expenses"

    id = Column(Integer, primary_key=True, index=True)
    store_id = Column(String(100), nullable=False)                   # 매장 식별 아이디
    amount = Column(Integer, nullable=False, default=0)              # 지출 금액 (원)
    category = Column(String(50), nullable=False)                    # 지출 카테고리 (예: 원두매입, 소모품비 등)
    description = Column(String(255), nullable=True)                 # 상세 설명
    expense_date = Column(Date, nullable=False)                      # 지출 일자
    created_at = Column(DateTime, nullable=False, server_default=func.now())


# 4. 특정 기간 동안 계산된 직원의 예상 급여 정보를 저장하는 테이블입니다.
class EstimatedPayroll(Base):
    """예상 급여 결과 저장 모델"""
    __tablename__ = "estimated_payrolls"

    id = Column(Integer, primary_key=True, index=True)
    # 어떤 직원의 급여인지 설정 (외래키 연결)
    employee_id = Column(Integer, ForeignKey("employees.id", ondelete="CASCADE"), nullable=False)
    period_start = Column(String(20), nullable=False)   # 정산 시작일 (YYYY-MM-DD)
    period_end = Column(String(20), nullable=False)     # 정산 종료일 (YYYY-MM-DD)
    total_work_hours = Column(Float, nullable=False)    # 총 근무 시간
    estimated_salary = Column(Integer, nullable=False)  # 계산된 예상 급여액 (원)
    calculated_at = Column(DateTime, nullable=False, server_default=func.now())  # 계산이 실행된 시각


# 5. 특정 기간 동안의 매출, 지출, 인건비를 종합한 예상 정산 정보를 저장하는 테이블입니다.
class EstimatedSettlement(Base):
    """예상 정산 결과 저장 모델"""
    __tablename__ = "estimated_settlements"

    id = Column(Integer, primary_key=True, index=True)
    period_start = Column(String(20), nullable=False)   # 정산 시작일 (YYYY-MM-DD)
    period_end = Column(String(20), nullable=False)     # 정산 종료일 (YYYY-MM-DD)
    total_sales = Column(Integer, nullable=False)       # 해당 기간의 예상 매출액 (원)
    total_expense = Column(Integer, nullable=False)     # 해당 기간의 예상 매입/운영 비용 (원)
    total_payroll = Column(Integer, nullable=False)     # 해당 기간의 예상 총 인건비 (원)
    other_expense = Column(Integer, nullable=False, default=0) # 해당 기간의 기타 추가 지출 비용 (원)
    net_profit = Column(Integer, nullable=False)        # 예상 매출 - (비용 + 인건비 + 기타비용) 순수익 (원)
    calculated_at = Column(DateTime, nullable=False, server_default=func.now())  # 계산이 실행된 시각


# 6. 직원별 기피/불가 시간 설정을 저장하는 테이블입니다. (신규 추가)
class EmployeeUnavailability(Base):
    """직원 기피/불가 시간 설정 모델"""
    __tablename__ = "employee_unavailabilities"

    id = Column(Integer, primary_key=True, index=True)
    # 기피 시간을 신청한 직원의 외래키 아이디
    employee_id = Column(Integer, ForeignKey("employees.id", ondelete="CASCADE"), nullable=False)
    
    # 기피 유형: 'weekly_recurring' (매주 특정 요일 반복) 또는 'specific_date' (특정 날짜 지정)
    unavailability_type = Column(String(30), nullable=False, default="weekly_recurring")
    
    # 요일 반복일 때 요일 번호 (0=월요일, 1=화요일, ..., 6=일요일)
    day_of_week = Column(Integer, nullable=True)
    
    # 특정 날짜 지정일 때 날짜 문자열 (YYYY-MM-DD 포맷)
    specific_date = Column(String(10), nullable=True)
    
    # 시작 시각 (0 ~ 23시)
    start_hour = Column(Integer, nullable=False, default=0)
    
    # 종료 시각 (1 ~ 24시)
    end_hour = Column(Integer, nullable=False, default=24)
    
    # 제약 수준: 'hard' (절대 근무 불가 - 배정 금지) | 'soft' (가급적 회피 - 페널티 부여 후 가급적 배정 지양)
    restriction_level = Column(String(10), nullable=False, default="hard")
    
    # 기피 신청 사유 (예: 학원 수업, 병원 방문 등)
    reason = Column(String(255), nullable=True)
    
    # 생성 일시
    created_at = Column(DateTime, nullable=False, server_default=func.now())

